Pass ddp buck cap size list additionally

NatureLLL opened the pull request:
Summary:
Pass `ddp_bucket_cap_mb_list` parameter to enable fine-grained control over DDP backward all_reduce bucket sizes, improving overlap between communication and computation.

This extends the existing single `ddp_bucket_cap_mb` parameter by supporting a list of bucket sizes for more precise tuning. Design details: https://docs.google.com/document/d/12J3RMSd4_SQHMIgU0Iw2fzfd_KfG4c9SDmFKQbnq9Zs/edit?usp=sharing
---
> Generated by [Confucius Code Assist (CCA)](https://www.internalfb.com/wiki/Confucius/Analect/Shared_Analects/Confucius_Code_Assist_(CCA)/)
[Session](https://www.internalfb.com/confucius?session_id=a9220f08-c678-11f0-8862-5fc1c72882bb&tab=Chat), [Trace](https://www.internalfb.com/confucius?session_id=a9220f08-c678-11f0-8862-5fc1c72882bb&tab=Trace)

Test Plan:
Verified that all_reduce operations fully overlap with computation when using 2D sharding with custom bucket size list.

Testing details and results: https://docs.google.com/document/d/12J3RMSd4_SQHMIgU0Iw2fzfd_KfG4c9SDmFKQbnq9Zs/edit?tab=t.0#bookmark=id.x6yllqdi4k1v
---
Test with HIM IG_CTR model,
[trace](https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree%2Ftraces%2Fdynocli%2Faps-aps-train-833023050-f833035934%2F0%2Frank-2.Nov_22_13_13_16.6312.pt.trace.json.gz&bucket=aps_traces) now
 {F1983770740} 
f833023050
---
[trace](https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree%2Ftraces%2Fdynocli%2Faps-aps-train-818234968-f818244217%2F0%2Frank-2.Oct_23_16_39_14.5677.pt.trace.json.gz&bucket=aps_traces) before
 {F1983770694} 
 f818234968

Differential Revision: D85873219




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

jbschlosser commented:
@wconstab are you okay to review this or can you please suggest someone else from the distributed side to do so?

mori360 commented:
Based on the traces, there's less forward but longer backward, is it expected?

NatureLLL committed:
Pass ddp buck cap size list additionally (#169026)

Summary:
Pass `ddp_bucket_cap_mb_list` parameter to enable fine-grained control over DDP backward all_reduce bucket sizes, improving overlap between communication and computation.

This extends the existing single `ddp_bucket_cap_mb` parameter by supporting a list of bucket sizes for more precise tuning. Design details: https://docs.google.com/document/d/12J3RMSd4_SQHMIgU0Iw2fzfd_KfG4c9SDmFKQbnq9Zs/edit?usp=sharing
---
> Generated by [Confucius Code Assist (CCA)](https://www.internalfb.com/wiki/Confucius/Analect/Shared_Analects/Confucius_Code_Assist_(CCA)/)
[Session](https://www.internalfb.com/confucius?session_id=a9220f08-c678-11f0-8862-5fc1c72882bb&tab=Chat), [Trace](https://www.internalfb.com/confucius?session_id=a9220f08-c678-11f0-8862-5fc1c72882bb&tab=Trace)


Test Plan:
Verified that all_reduce operations fully overlap with computation when using 2D sharding with custom bucket size list.

Testing details and results: https://docs.google.com/document/d/12J3RMSd4_SQHMIgU0Iw2fzfd_KfG4c9SDmFKQbnq9Zs/edit?tab=t.0#bookmark=id.x6yllqdi4k1v
---
Test with HIM IG_CTR model,
[trace](https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree%2Ftraces%2Fdynocli%2Faps-aps-train-833023050-f833035934%2F0%2Frank-2.Nov_22_13_13_16.6312.pt.trace.json.gz&bucket=aps_traces) now
 {F1983770740} 
f833023050
---
[trace](https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree%2Ftraces%2Fdynocli%2Faps-aps-train-818234968-f818244217%2F0%2Frank-2.Oct_23_16_39_14.5677.pt.trace.json.gz&bucket=aps_traces) before
 {F1983770694} 
 f818234968





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

Reviewed By: Fiery

Differential Revision: D85873219

Pulled By: NatureLLL
